### PR TITLE
Gimbal control

### PR DESCRIPTION
**Pull request by YusukeTakase-Sensyn**
control gimbal, camera, camera param by action

### CODE REVIEW

File: include/action_type_definition.h [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    TakeOff,
+    AimCamera,
+    TargetCamera,
+    TakePicture,
+    RecordVideo,
+    StartRecordVideo,
+    StopRecordVideo
+};
+
+std::map<int, std::string> map_GimbalCameraAction = {
```
Comment (MasahiroOgawa) [inline include/action_type_definition.h:21]: Why do you need to define this? You already defined enum class, so you can use it by the name.

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    tf2
+    tf2_ros
+    tf2_geometry_msgs
+    tf
+    pcl_ros
+    actionlib_msgs
+    dji_osdk_ros
+    )
+
+find_package(PCL REQUIRED)
```
Comment (MasahiroOgawa) [inline CMakeLists.txt:24]: Why do you need PCL for gimbal control?
If this include some recognition module, it should be separeted.

File: include/dji_gimbal_camera_action_client.h [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+#include <dji_osdk_ros/dji_vehicle_node.h>
+#include <sensyn_dji_gimbal_action/GimbalCameraAction.h>
+
+#include <action_type_definition.h>
+
+using namespace actionlib;
+using namespace sensyn_dji_gimbal_action;
+using namespace std;
+
+namespace sr {
```
Comment (MasahiroOgawa) [inline include/dji_gimbal_camera_action_client.h:19]: Do we always use namespace sr?
I think it is clearer to use sensyn.

File: include/dji_gimbal_camera_action_client.h [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+    class GimbalCameraActionClient : public SimpleActionClient<sensyn_dji_gimbal_action::GimbalCameraAction> {
+        using SimpleActionClient::SimpleActionClient;
+
+    public:
+        void doneCb(const actionlib::SimpleClientGoalState &state, GimbalCameraResult &result);
+
+        bool requestTaskAimCamera(const double roll, const double pitch, const double yaw, const double zoom) ;
```
Comment (MasahiroOgawa) [inline include/dji_gimbal_camera_action_client.h:27]: I you use roll pitch yaw as an angle representation, you need to specify the order explicitly.
If you use the same order of DJI definition, you should write the link to the doc in the code.

File: src/dji_camera_action_server.cpp [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+      if (video_action_client_.call(camera_record_video_action)) {
+          ROS_INFO_STREAM(" record success!");
+
+          result_shootCamera.done = true;
+      } else {
+          ROS_ERROR_STREAM(" record failed!");
+          result_shootCamera.done = false;
+      }
+      ros::Duration(2.0).sleep();
```
Comment (MasahiroOgawa) [inline src/dji_camera_action_server.cpp:46]: You should create the variable for this 2.0 [sec].

File: src/dji_gimbal_camera_action_server.cpp [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+
+      result = true;
+  } else {
+      ROS_ERROR_STREAM(" gimbal_action reset failed!");
+      result = false;
+      return result;
+  }
+
+  //rotate gimbal
+
```
Comment (MasahiroOgawa) [inline src/dji_gimbal_camera_action_server.cpp:92]: What happen if the delay is few second?
Can you suppress the vibration?
If no, it's better where we should update the gimbal control.

**Review by MasahiroOgawa — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 2 above)
Comment (ohtt) [reply CMakeLists.txt:24]: pclは消したほうがいいかな

**Pull request closed without merging**